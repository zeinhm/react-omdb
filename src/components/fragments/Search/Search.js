import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'
import './styles.scoped.css'

export default function Search(props) {
  const history = useHistory()
  const { list, keyword, onChange, onClick, onSubmit } = props

  const [activeOption, setActiveOption] = useState(false)
  const [chosenOption, setChosenOption] = useState('')
  const [options, setOptions] = useState([])
  
  const state = useSelector(state => state.home)
  
  useEffect(() => {
    if (list && keyword) setOptions(list)
  }, [list])

  useEffect(() => {
    if (!keyword) setOptions([])
  }, [keyword])

  useEffect(() => {
    if (chosenOption) history.push(`/detail/?id=${chosenOption}`)
  },[chosenOption])

  const onClickSearch = () => {
    setOptions([])
    onClick()
  }
  const onClickOption = id => {
    setOptions([])
    setChosenOption(id)
  }
  const onKeyDown = e => {
    if (e.keyCode === 13) {
      if (activeOption === false) {
        setOptions([])
        onSubmit()
      } else if(activeOption === 5) {
        setOptions([])
        onClickSearch()
      } else {
        setActiveOption(false)
        setOptions([])
        setChosenOption(options[activeOption].imdbID)
      }
    } else if (e.keyCode === 38) {
      if (activeOption === 0) {
        setActiveOption(false)
      } else {
        setActiveOption(activeOption - 1)
      }
    } else if (e.keyCode === 40) {
      if (activeOption === false) {
        setActiveOption(0)
      } else {
        if (activeOption === options.length - 1) {
          return;
        }
        setActiveOption(activeOption +1 )
      }
    } else if (e.keyCode === 27) {
      setActiveOption(false)
      setOptions([])
    }
  }

  const optionList = () => {
    return (
      <ul className="options">
        {options?.length >= 1 && options.map((i, idx) => {
          let className = ''
          if (idx === activeOption) {
            className = 'active'
          }
          if (idx < 5) {
            return <li className={className} key={idx} onClick={() => onClickOption(i.imdbID)}>{i.Title}</li>
          }
        })}
        {!state.messageAutoComplete && options.length > 5 && <li className={activeOption === 5 ? 'active' :''} onClick={() => onClickSearch()}>... <span>&#x2192;</span></li>}
        {(state.messageAutoComplete && keyword.trim() !== '' && !state.isLoadingGetMovies) && (
          <li className='error'>{state.messageAutoComplete}</li>
        )}
      </ul>
    )
  }
  
  return (
    <div className='search-box'>
        <input className='search-input' onChange={e=>onChange(e.target)} onKeyDown={(e)=> onKeyDown(e)}
          placeholder='Search...' type="text" value={keyword}  />
        <input className={`search-btn ${state?.isLoadingAutoComplete ? 'spinner-active' : 'spinner-inactive'}`}
          onClick={()=>onClickSearch()}  type="submit"/>    
      {optionList()}
    </div>
  )
}

Search.defaultProps = {
  list: [],
  keyword: '',
  onChange: () => {},
  onClick: () => {},
  onSubmit: () => {}
}

Search.propTypes = {
  list: PropTypes.array,
  keyword: PropTypes.string,
  onChange: PropTypes.func,
  onClick: PropTypes.func,
  onSubmit: PropTypes.func
}
