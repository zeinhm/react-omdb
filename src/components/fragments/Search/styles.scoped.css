.search-box {
  margin: 0 auto;
  position: relative;
  width: 20rem;
}
.search-input {
  border: 4px solid transparent;
  border-radius: 10px 0 ;
  box-shadow: 0.6875rem 0.5625rem 1.3125rem rgba(173, 182, 217, 0.5);
  font-size: 1rem;
  padding: 1rem;
  transition: width 0.3s;
  width: 100%;
}
.search-input:focus {
  border: 1px solid #08a1b6;
  outline: none;
  width: 100%;
}
.search-btn {
  background-color: transparent;
  background-repeat: no-repeat;
  background-size: 100%;
  border: 0;
  height: 100%;
  opacity: 0.2;
  position: absolute;
  right: 1rem;
  text-indent:-999px;
  top: 50%;
  transform: translateY(-20%);
  width: 1.5rem;
}

.search-btn.spinner-inactive {
  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAQAAAAAYLlVAAACnElEQVR4AcXZsUsbYRjH8e+dh2s1SyAGJwMJuDj1BIcEhJQIOnTq5F+QOf0jIq79A7oFh7aYyVBEkaZDC3awECc1AUXRIqUQotfFocnjJe/dk+b9PKP65Md7z13ee3Uwk2SNHKmngs5TnbDLJQqjA+RYZ4OXuDzvkSYf+cAJ44fPAYFhHeAzVhlqBBGrRoax8KjSJYhRXap4KCVoECiqQQKFLC0CZbXIElOBOwJ9cUchzm2Y5QsveN4tdfY4o00HSDHHPKuUmOV5v/D5SSSJ0MXfIY+HBB55dkIvRIIIvJDR28dnFJ/9kHH0MFaVDehRxlSZnuxAFUMZunKQKBJFUQ4wXTIYqcmPZ5GoFmUEahjw5eJTJI6ivBD4jCS/csrEVZZfU4yQk5OPhrwjcoRygQ0GVdCQf73OUEfisaMkHk1HDJHkYeDX82jlBzo+kCSEyxruwDP/EK1DbsWnhHDFgNTpodWjLgY9NECKfnvoyS4p8wBngN5Z/ABtQK8dP0AH0OuYB5iMqfAAMque7HJtHmAOPdnlxjzAPHqyy5V5gFX0ZJfj8AAn9CvhoeVRol8zPMAuj/xrlhW0Vpg1D3ApflhGR3b4wTlDvI24i4u+w9y0uyVrM213U1qxuy2/Z8bui8m23VezgGW7L6cBLdIWXs9FBAsHFCLCJI9opFMKXEzkkEp/IbK0bEdI0LARQRzVWoigPKy+Z5tlWooIiuP6NhVmAEiPNwLkqHDEw5CGx2wyDQDRI8T7l80U19xwxTFNmpwzKM1nFsyeCw7jFymCAxYjrHDp8r9cUOCUYRZ4Bw6AxVV47QJYXIVXLliNsOSC1Qh/XLAa4ZuDmmIcH1l2AaytwhZfmaAkn/qOb7eYBofJekOJJX7znfccAvwFyB3OeNys7d4AAAAASUVORK5CYII=");
}

.search-btn.spinner-active {
  background-image: url('../../../assets/spinner.svg');
}

.search-btn:hover {
  cursor: pointer;
  opacity: 0.4;
  outline: none;
}
.search-btn:focus {
  opacity: 0.6;
  outline: none;
}

ul.options {
  background: white;
  display: block;
  list-style: none;
  transition: width 0.3s;
  margin: auto;
  position: absolute;
  top: 100%;
  width: 20rem;
}

ul.options li {
  background: white;
  border-radius: 2px;
  box-shadow: -0.75rem 0.5rem 1.25rem rgba(173, 182, 217, 0.170884);
  display: block;
  margin: 10px auto;
  font-size: 1rem;
  padding: 10px;
  width: 100%;
}

ul.options li:nth-last-child(1) {
  margin-bottom: 0;
  position: relative;
}

ul.options li:nth-last-child(1) span {
  position: absolute;
  right: 1rem;
}

ul.options li:not(.error):hover {
  cursor: pointer;
  font-size: 1.125rem;
  transition: 0.1s all;
}

ul.options li.active {
  background: rgba(173, 182, 217, 0.25);
  font-size: 1.125rem;
}

ul.options li.error {
  font-style: italic;
  color: lightslategray;
}

.spinner {
  position: absolute;
  right: 5%;
  top: 50%;
  transform: translateY(-25%);
  width: fit-content;
}
