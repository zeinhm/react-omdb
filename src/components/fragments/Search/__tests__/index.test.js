import React from 'react';
import { useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Search from '../Search';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(cb => {
    cb({});
    return {
      state: {
        isLoadingAutoComplete: false
      }
    };
  }),
}));

jest.mock('react-router-dom', () => ({
  useHistory: () => ({ push: jest.fn() }),
}));

describe('src/components/fragments/Search', ()=> {
  test('Search', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Search />);
    useSelector.mockImplementationOnce(() => ({
      state: {
        isLoadingAutoComplete: true
      }
    }));
    expect(tree).toMatchSnapshot();
  });
});
