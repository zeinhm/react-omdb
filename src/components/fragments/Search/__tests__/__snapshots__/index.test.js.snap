// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/Search Search 1`] = `
<div
  className="search-box"
>
  <input
    className="search-input"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search..."
    type="text"
    value=""
  />
  <input
    className="search-btn spinner-inactive"
    onClick={[Function]}
    type="submit"
  />
</div>
`;
